### PR TITLE
Use Salt for mach bootstrap

### PR DESCRIPTION
**Pull request by aneeshusa**
This is currently WIP, but allows Salt for `mach bootstrap`. Not looking for review yet, just posting for visibility. You can run `./mach bootstrap` and Salt will run, letting you know what changes it would make for bootstrapping (doesn't actually run yet though).

Currently, this reads from saltfs using gitfs, meaning it always tracks the master branch. (Note that this is blocked on https://github.com/servo/saltfs/pull/577 landing; in the meantime, I've included a small workaround in this PR to pull from my updated saltfs branch, which will need to be removed.) In the future, the relevant Salt code may be merged in tree as part of Docker support for TC, and the bootstrapper should be updated to read from in tree.

Also, our Windows machines have not been Salted, so the existing Windows bootstrappers are retained. 

TODO:
- [x] Hook into existing bootstrapping code less hackily
- [x] Actually bootstrap instead of always using `test=True` mode (includes sudo)
- [x] Default to interactive mode (test first, then ask), with a force flag
- [x] Don't require running from the repository root directory
- [x] Make it easy to add support for other distros 

---

- [ ] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because they should be verified manually

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14974)

### CODE REVIEW

**Comment by aneeshusa** [comment]
This should work on Ubuntu Trusty now.

@Wafflespeanut @frewsxcv Python question: So far it seems necessary to use `--system-site-packages` when creating the virtualenv to allow Salt to work properly (see the commit message for details). Does this seem OK to you or should I make a separate virtualenv for Salt?

**Comment by aneeshusa** [comment]
I've finished the remaining items on this, including a cleanup/overhaul of the existing bootstrapping. This is ready for review.

cc @UK992 since you wrote the original Windows bootstrapping, can you check that `mach bootstrap` still works on Windows?

**Comment by aneeshusa** [comment]
Note this is still blocked on servo/saltfs#577 landing, however.

File: python/servo/bootstrap.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        # so only detect failures of the actual salt-call binary itself.
+        retcode = subprocess.call(cmd + ['test=True'])
+        if retcode != 0:
+            print('Something went wrong while bootstrapping')
+            return retcode
+
+        proceed = raw_input(
+            'Proposed changes are above, proceed with bootstrap? [y/N]: '
+        )
+        if proceed not in ['y', 'yes', 'Y', 'Yes', 'YES']:
```
Comment (Wafflespeanut) [inline python/servo/bootstrap.py]: `if proceed.lower() not in ['y', 'yes']`? :smile:

**Comment by Wafflespeanut** [comment]
This is great! :smile: 

> Does this seem OK to you or should I make a separate virtualenv for Salt?

This should be fine, since `pip install -I` is going to reinstall our local requirements which will shadow the globally available libraries anyway.

**Comment by aneeshusa** [comment]
Also, I think this should work on macOS (using Homebrew) - if someone could check that I'll enable the Salt bootstrapper for macOS as well (instead of just Ubuntu 14.04).

**Comment by UK992** [comment]
Since both gnu and msvc use same python, detection should be relied on `host_triple`.
For msvc, maybe could `--force` flag delete `msvc-dependencies` folder and reinstall all packages?
Also salt should be excluded  in `requirements.txt` for Windows, because it's failed to create virtualenv.

**Comment by UK992** [comment]
Also `set()` on `WINDOWS_GNU` packages list should be removed, because it fails to execute command.

**Comment by aneeshusa** [comment]
@UK992 
- I haven't changed the GNU/MSVC detection from the original. I'm not sure exactly how to use `host_triple`, can you show me how I should use it here?
- We shouldn't never need to delete the `msvc-dependencies` folder, since each package version is installed in its folder and never changed, so IMO a `--force` flag is unnecessary. If we do want that behavior it's different from what `--force` does for the other two, so I think we would need two flags at that point.
- Addressed your other two comments.

**Comment by UK992** [comment]
GNU/MSVC detection was broken for some time. Using `host_triple`:
```python
if "windows-gnu" in host_triple():
    bootstrapper = windows_gnu
elif "windows-msvc" in host_triple():
    bootstrapper = windows_msvc
```

**Comment by aneeshusa** [comment]
@UK992 OK thanks, I updated that. Can you give it a try and also run `./mach build` on MSVC after bootstrapping? I want to ensure the reorganizing of `msvc-dependencies` didn't break the `build_env` function.

**Comment by UK992** [comment]
@aneeshusa building works perfectly on MSVC.

**Comment by aneeshusa** [comment]
servo/saltfs#577 merged and I've updated this, so this is no longer blocked.

**Comment by jdm** [comment]
@metajack Review ping!

**Comment by metajack** [comment]
@bors-servo r+

Looks good to me.

**Pull request merged**